LinGui: add support for running a command after an encode completes

samhutchins opened the pull request:
**Description of Change:**

My lack of C and glib/gtk experience may very well be showing in full force here...

macOS and Windows can send the encoded file to an external application after the encode completes, this change adds that functionality for the Linux GUI as well. I'm not sure if this is the best way to invoke an external command

**Tested on:**

- [ ] Windows 10+  (via MinGW)
- [ ] macOS 10.13+
- [x] Fedora Linux

samhutchins committed:
LinGui: add support for running a command after an encode completes

robxnano commented:
Looks good, but have you considered making this more useful in the Flatpak version? Users will probably expect the command to run outside the sandbox. You could do this by prepending `flatpak-spawn --host -- ` to the argv of the command when running in the Flatpak environment, which you can detect with `if (g_access("/.flatpak-info", F_OK) == 0)`. Note that this would also need permission to talk to the org.freedesktop.Flatpak D-Bus name.

samhutchins commented:
I had not, I've never really used the flatpak version. Thanks for the pointers on where to start, I'll look into it

samhutchins committed:
Detect Flatpak and prepend flatpak-spawn --host

samhutchins commented:
I've made the changes I _think_ are needed, but I don't seem to be able to launch vlc when running a flatpack version of HandBrake. Is there something I'm missing with permissions?

sr55 commented:
@robxnano or @jstebbins  Any advice if you have a minute

jstebbins commented:
flatpaks don't have direct access to the hosts applications.  I.e. if you run
```
flatpak run --command=bash fr.handbrake.ghb
ls /usr/bin
```
you will see a different list of applications than is available in /usr/bin outside of the flatpak.

The flatpak ecosystem provides portals for things like this, but I don't know if there is a portal for execution of arbitrary apps. There is however a portal for launching the default app for a uri, `g_app_info_launch_default_for_uri()`
See https://docs.flatpak.org/en/latest/portals.html and https://blogs.gnome.org/mclasen/2016/07/08/portals-using-gtk-in-a-flatpak/

*Maybe* `g_app_info_launch()` also has this ability to go through the portal, but I doubt it. I could not find any information indicating this would work.

samhutchins commented:
Apologies for dropping the ball with this one, life and all that. I have worked out how to get this to work with flatpak though. I had `--system-talk-name` instead of `--talk-name` for the `org.freedesktop.Flatpak` permission. With that corrected, `flatpak spawn --host --` does what it should

I've rebased and fixed the flatpak permissions

I'm not expecting this PR to be merged for 1.9, as it does contain a new string in the Linux UI

sr55 commented:
Since we are well past 1.9, is this in a state to consider merging?

samhutchins commented:
I think so, I've addressed all the feedback I've received so far

sr55 commented:
OK, Thank you!
